lib/ThreadPool.h

```diff
@@ -0,0 +1,184 @@
+/*
+ * ThreadPool.h, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#pragma once
+
+#include <boost/thread/future.hpp>
+#include <boost/thread/condition_variable.hpp>
+
+#include "ThreadUtilities.h"
+
+VCMI_LIB_NAMESPACE_BEGIN
+
+//Credit to https://github.com/Liam0205/toy-threadpool/tree/master/yuuki
+
+template <typename T>
+class BlockingQueue : boost::noncopyable
+{
+public:
+	template <typename... Args>
+	void emplace(Args&&... args)
+	{
+		boost::mutex::scoped_lock lock(mx);
+		queue.emplace(std::forward<Args>(args)...);
+	}
+
+	bool pop(T& holder)
+	{
+		boost::mutex::scoped_lock lock(mx);
+		if (queue.empty())
+			return false;
+
+		holder = std::move(queue.front());
+		queue.pop();
+		return true;
+	}
+
+private:
+	std::queue<T> queue;
+	mutable boost::mutex mx;
+};
+
+class TaskGroup
+{
+	friend class ThreadPool;
+
+	std::vector<boost::future<void>> tasks;
+	std::atomic<int> tasksLeft = 0;
+};
+
+class ThreadPool
+{
+	friend class TaskGroup;
+private:
+	using Functor = std::function<void()>;
+	mutable boost::mutex mx;
+	mutable boost::condition_variable_any cv;
+
+	std::atomic<bool> stopping = false;
+
+	std::vector<boost::thread> workers;
+	mutable BlockingQueue<Functor> tasks;
+
+	void spawn();
+	void terminate();
+
+public:
+	static ThreadPool & global();
+
+	explicit ThreadPool();
+	~ThreadPool();
+
+	void addTask(TaskGroup & group, std::function<void()>&& f);
+	void waitFor(TaskGroup & group);
+};
+
+inline ThreadPool & ThreadPool::global()
+{
+	static ThreadPool globalInstance;
```

Comment:
Are you sure we should keep static instance in header. Also we have TBB anyway with its own thread pool. So maybe it is less convenient to use it, I do not know

Reply:
Probably TBB is more complex.

Reply:
Static inside inline function should work as intended, so that part is OK.

Thread pool from TBB - will look into it.
If they have persistent, preconfigured thread pool - then probably makes sense to use it. But probably an overkill for our needs.

I am aiming for minimal overhead / minimal configuration to a point where you can parallelize operation that takes 1-2 ms. Meanwhile TBB:
>  CAUTION:
Typically a loop needs to take at least a million clock cycles to make it worth using parallel_for. For example, a loop that takes at least 500 microseconds on a 2 GHz processor might benefit from parallel_for.